### PR TITLE
Add time quota in checking global objects

### PR DESCRIPTION
**Pull request by liufengyun**
Make global object checker default to best-effort mode

- In best-effort mode, the checker tries to be fast, useful and unobtrusive.
- In the aggressive mode, the checker tries to be sound by spending more check time and produce more verbose warnings.

In both modes, there will be a worst-case performance guarantee based on a quota on the number of method calls in initializing a global object.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/init/Objects.scala [hunk 1]
```diff
@@ -362,18 +422,35 @@ class Objects(using Context @constructorOnly):
...
         log("Iteration " + count) {
-          data.checkingObjects += obj
-          init(tpl, obj, classSym)
+          data.addChecking(obj)
+
+          try
+            init(tpl, obj, classSym)
+          catch case _: OutOfQuotaException =>
+            if !BestEffort then
+              report.warning("Giving up checking initializatino of " + classSym + " due to exhausted budget", classSym.sourcePos)
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/transform/init/Objects.scala]: Should we turn on this warning for best-effort mode?

Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/init/Objects.scala]: Discussed this in meeting and it's agreed useful to display the warning in both modes: end users can learn from the warning that their initialization code is too complex.

**Comment by liufengyun** [comment]
It's interesting to see that the changes do not impact the test cases. With the assumption that most real-world cases are simple, we expect the changes to NOT impact most use cases, but only defend against edge cases.

File: compiler/src/dotty/tools/dotc/transform/init/Objects.scala [hunk 2]
```diff
@@ -362,18 +422,34 @@ class Objects(using Context @constructorOnly):
...
         given Scope = obj
         log("Iteration " + count) {
-          data.checkingObjects += obj
-          init(tpl, obj, classSym)
+          data.addChecking(obj)
+
+          try
+            init(tpl, obj, classSym)
+          catch case _: OutOfQuotaException =>
+            report.warning("Giving up checking initializatino of " + classSym + " due to exhausted budget", classSym.sourcePos)
```
Comment (olhotak) [inline compiler/src/dotty/tools/dotc/transform/init/Objects.scala]:
```suggestion
            report.warning("Giving up checking initialization of " + classSym + " due to exhausted budget", classSym.sourcePos)
```

File: compiler/src/dotty/tools/dotc/transform/init/Objects.scala [hunk 3]
```diff
@@ -362,18 +422,34 @@ class Objects(using Context @constructorOnly):
...
         val hasError = ctx.reporter.pendingMessages.nonEmpty
         if cache.hasChanged && !hasError then
-          cache.prepareForNextIteration()
-          iterate()
+          if count <= 3 then
+            cache.prepareForNextIteration()
+            iterate()
+          else
+            if !BestEffort then
+              report.warning("Giving up checking initializatino of " + classSym + " due to iteration > = " + count, classSym.sourcePos)
```
Comment (olhotak) [inline compiler/src/dotty/tools/dotc/transform/init/Objects.scala]:
```suggestion
              report.warning("Giving up checking initialization of " + classSym + " due to iteration > = " + count, classSym.sourcePos)
```

File: compiler/src/dotty/tools/dotc/transform/init/Objects.scala (hunk 1 above)
Comment (eejbyfeldt) [reply compiler/src/dotty/tools/dotc/transform/init/Objects.scala]: The error message could be improved if it included suggestions to the user what one can do to address the issue.

Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/init/Objects.scala]: Thank you for the feedback @eejbyfeldt , that is a good suggestion. Currently, we are still experimenting with it. We are not sure about what interface to provide to improve user experience:

- An option to skip the object
- An option to increase the budge
- To be best effort and suppress the warnings?

It is still too early to make a decision. Eventually we will provide a design based the feedback from community.
